### PR TITLE
to_chars integer parser

### PR DESCRIPTION
**Pull request by mborland**

### CODE REVIEW

**Comment by mborland** [comment]
@pdimov Do you see anything obvious as to why Windows is failing in CI? It's all the same linker error.

**Comment by pdimov** [comment]
14.0 because of this:
```
D:\a\charconv\boost-root\boost/charconv/to_chars.hpp(107): error C2220: warning treated as error - no 'object' file generated
D:\a\charconv\boost-root\boost/charconv/to_chars.hpp(107): warning C4127: conditional expression is constant
```

**Comment by pdimov** [comment]
The rest seems to be because they try to link to the DLL, but it exports nothing, so no LIB is generated.

I suppose now that the functions are constexpr, nothing remained in the DLL. But there will be one day, so maybe add a stub fp function just to have something exported.

**Comment by mborland** [comment]
@pdimov Can you take a look through this? I think it is in good shape.

File: test/Jamfile [hunk 1]
```diff
@@ -21,3 +22,4 @@ run from_chars.cpp ;
 run to_chars.cpp ;
 run roundtrip.cpp ;
 run from_chars_STL_comp.cpp : : : [ requires cxx17_std_apply ] ;
+run to_chars_integer_STL_comp.cpp : : : [ requires cxx17_std_apply ] ;
```
Comment (pdimov) [inline test/Jamfile]: cxx17_std_apply doesn't seem the right condition here; why not cxx17_hdr_charconv?

File: include/boost/charconv/detail/integer_conversion.hpp [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+namespace boost { namespace charconv { namespace detail {
+
+constexpr std::uint64_t pack(std::uint32_t word1, std::uint32_t word2) noexcept
+{
+    return static_cast<std::uint64_t>(word1) << 32 | word2;
+}
+
+BOOST_CXX14_CONSTEXPR std::pair<std::uint32_t, std::uint32_t> unpack(std::uint64_t value)
+{
+    auto x = static_cast<std::uint32_t>((value & UINT64_MAX) >> UINT64_C(32));
```
Comment (pdimov) [inline include/boost/charconv/detail/integer_conversion.hpp]: `& UINT64_MAX` is a no-op here.

**Comment by pdimov** [comment]
Looks good overall. The html should be in doc/html though, not doc, and shouldn't be checked in.

I think we need significantly more test values than that; esp. against `<charconv>`, we need to be testing with both hand-picked values (1, 9, 10, 99, 100, 999, etc) and random values (e.g. from splitmix64.)

(Like https://github.com/boostorg/core/blob/8977da6f50b3a29e429c9f78766296f66e2dab26/test/bit_ceil_test.cpp#L51-L62 for instance, except maybe with more than 1000 values.)

File: include/boost/charconv/detail/is_constant_evaluated.hpp [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+//  LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef BOOST_CHARCONV_TOOLS_IS_CONSTANT_EVALUATED_HPP
+#define BOOST_CHARCONV_TOOLS_IS_CONSTANT_EVALUATED_HPP
+
+#include <boost/charconv/config.hpp>
+
+#ifdef __has_include
+# if __has_include(<version>)
+#  include <version>
```
Comment (pdimov) [inline include/boost/charconv/detail/is_constant_evaluated.hpp]: We don't need `<version>` here. Just include `<type_traits>` and check the macro, without any of that has_include business.

File: include/boost/charconv/detail/memcpy.hpp [hunk 4]
```diff
@@ -9,6 +9,19 @@
...
+// ‘void* __builtin___memcpy_chk(void*, const void*, long unsigned int, long unsigned int)’ specified size between 
+// 18446744071562067968 and 18446744073709551615 exceeds maximum object size 9223372036854775807 [-Werror=stringop-overflow=]
+//
+// memcpy is defined as taking a size_t for the count and the largest count this will recieve is the number of digits
+// in a 128-bit int (39) so we can safely ignore
+#if __GNUC__ >= 10
+#  pragma GCC diagnostic push
+#  pragma GCC diagnostic ignored "-Wstringop-overflow"
+#  define BOOST_CHARCONV_STRINGOP_OVERFLOW_DISABLED
+#endif
```
Comment (mborland) [inline include/boost/charconv/detail/memcpy.hpp:24]: @pdimov Any idea why GCC-11 would be still failing from "-Wstringop-overflow" with this? GCC-10 and 12 both had the same error from the same lines which were fixed ([old run here](https://github.com/CPPAlliance/charconv/actions/runs/4086466481)).

Comment (pdimov) [reply include/boost/charconv/detail/memcpy.hpp:24]: I don't know, sorry.

**Comment by sdarwin** [comment]
https://master.charconv.cpp.al/
https://develop.charconv.cpp.al/

File: include/boost/charconv/detail/memcpy.hpp (hunk 4 above)
Comment (mborland) [reply include/boost/charconv/detail/memcpy.hpp:24]: Looks like it is related to this bug in GCC that was reported in 4.7.0 and will finally be fixed in 13 https://gcc.gnu.org/bugzilla/show_bug.cgi?id=53431

**Comment by mborland** [comment]
@pdimov Can you please take a final look at this? I think I addressed all of your previous comments.

**Comment by pdimov** [comment]
Looks good to merge.

**Pull request merged**